### ISSUE TITLE
accumulated gradients

### ISSUE THREAD

**Issue by FanChiMao** [bubbliiiing/yolov8-pytorch#30]
您好，想請教B導您關於 Accumulate gradients 相關的問題:

在 YOLOv8 官方程式中，在訓練期間有使用到累積梯度的技巧，利用較小的學習率分批處理累加梯度，再統一更新。
- YOLOv8 ([1](https://github.com/ultralytics/ultralytics/blob/263bfd1e93a0ebab898892501bdb3ae36aa4aa0b/ultralytics/engine/trainer.py#L270-L278), [2](https://github.com/ultralytics/ultralytics/blob/263bfd1e93a0ebab898892501bdb3ae36aa4aa0b/ultralytics/engine/trainer.py#L358C1-L361C39))
 
    ```
        # (self.args.nbs = 64, self.batch_size = 8)
        self.accumulate = max(round(self.args.nbs / self.batch_size), 1)  # accumulate loss before optimizing
        weight_decay = self.args.weight_decay * self.batch_size * self.accumulate / self.args.nbs  # scale weight_decay
        iterations = math.ceil(len(self.train_loader.dataset) / max(self.batch_size, self.args.nbs)) * self.epochs
        self.optimizer = self.build_optimizer(model=self.model,
                                            name=self.args.optimizer,
                                            lr=self.args.lr0,
                                            momentum=self.args.momentum,
                                            decay=weight_decay,
                                            iterations=iterations)
        # ======================================================================
        # Optimize - https://pytorch.org/docs/master/notes/amp_examples.html
        if ni - last_opt_step >= self.accumulate:
          self.optimizer_step()
          last_opt_step = ni
    ```

而在您的代碼中，我觀察到似乎僅包含縮小 learning rate 的部分，而並沒有到達累積 step 後，才重新計算損失的程式。
- YOLOv8_bubbliiiing ([1](https://github.com/bubbliiiing/yolov8-pytorch/blob/c245eb0160316f101fc49d005068db495365ae5c/train.py#L428-L432))  

    ```
        nbs             = 64
        lr_limit_max    = 1e-3 if optimizer_type == 'adam' else 5e-2
        lr_limit_min    = 3e-4 if optimizer_type == 'adam' else 5e-4
        Init_lr_fit     = min(max(batch_size / nbs * Init_lr, lr_limit_min), lr_limit_max)
        Min_lr_fit      = min(max(batch_size / nbs * Min_lr, lr_limit_min * 1e-2), lr_limit_max * 1e-2)
   ```

故想請教您認為，是否有可能是因為這個原因導致模型訓練較為困難，需先凍結骨幹部分，後續在打開權重訓練，才較能避免模型梯度爆炸、消失等相關問題? 懇請您的回復，感謝您~
